File: XorShift.ts
"use strict";

export class XorShift{
	private r:Iterator<number>;
	public static defaults:{[s:string]:number}={
		x:123456789,
		y:362436069,
		z:521288629,
		w:88675123
	};
	public readonly seeds:{[s:string]:number};
	public randCount=0;

	constructor(w?:number,x?:number,y?:number,z?:number){
		if(typeof w=="undefined") w=0|Date.now();
		if(typeof x=="undefined") x=0|w<<13;
		if(typeof y=="undefined") y=0|(w>>>9)^(x<<6);
		if(typeof z=="undefined") z=0|y>>>7;
		this.seeds={x:x>>>0,y:y>>>0,z:z>>>0,w:w>>>0};
		this.randCount=0;
		this.r=this.randGen(w,x,y,z);
	}

	public *randGen(w:number,x:number,y:number,z:number):Iterator<number>{
		var t:number;
		for(;;){
			t=x^(x<<11);
			x=y;
			y=z;
			z=w;
			yield w=((w^(w>>>19))^(t^(t>>>8)))>>>0;
		}
	}

	public rand():number{
		this.randCount=0|this.randCount+1;
		return this.r.next().value;
	}

	public randInt(min=0,max=0x7FFFFFFF):number{
		return 0|this.rand()%(max+1-min)+min;
	}

	public randFloat(min=0,max=1):number{
		return Math.fround(this.rand()%0xFFFF/0xFFFF)*(max-min)+min;
	}

	public shuffle<T>(_arr:T[]):T[]{
		var arr=[..._arr];
		for(let i=0;i<=arr.length-2;i++){
			let r=this.randInt(i,arr.length-1);
			[arr[i],arr[r]]=[arr[r],arr[i]];
		}
		return arr;
	}

	public static defaultSeed=class extends XorShift{
		constructor(){
			super(
				XorShift.defaults["w"],
				XorShift.defaults["x"],
				XorShift.defaults["y"],
				XorShift.defaults["z"]
			);
		}
	}
}

File: XorShift.py
from typing import *
from time import time
from copy import copy

T=TypeVar("T")
class XorShift:
	defaults:Dict[str,int]={
		"x":123456789,
		"y":362436069,
		"z":521288629,
		"w":88675123
	}

	def __init__(self,w:int=None,x:int=None,y:int=None,z:int=None):
		if w is None: w=int(time())
		if x is None: x=w<<13
		if y is None: y=(w>>9)^(x<<6)
		if z is None: z=y>>7
		self.seeds:Dict[str,int]={"x":x,"y":y,"z":z,"w":w}
		self.randCount=0
		self.__r=self.randGen(w,x,y,z)

	def randGen(self,w:int,x:int,y:int,z:int)->Iterable[int]:
		t:int
		while True:
			t=x^(x<<11&0xFFFFFFFF)
			x,y,z=y,z,w
			w=(w^(w>>19))^(t^(t>>8))
			yield w

	def rand(self)->int:
		self.randCount+=1
		return next(self.__r)

	def randInt(self,min:int=0,max:int=0x7FFFFFFF)->int:
		return self.rand()%(max+1-min)+min

	def randFloat(self,min:float=0,max:float=1)->float:
		return self.rand()%0xFFFF/0xFFFF*(max-min)+min

	def shuffle(self,_arr:Union[List[T],Iterable[T]])->Union[List[T],Iterable[T]]:
		arr=copy(_arr)
		if isinstance(arr,list):
			isList=True
		elif isinstance(arr,range) or isinstance(arr,tuple):
			isList=True
			arr=list(arr)
		else:
			isList=False
			arr=list(arr)

		for i in range(0,len(arr)-1):
			r=self.randInt(i,len(arr)-1)
			arr[i],arr[r]=arr[r],arr[i]

		if not isList:
			arr=iter(arr)
		return arr

	def defaultSeed():return type("defaultSeed",(XorShift,),{
		"__init__":lambda self:
			XorShift.__init__(
				self,
				**XorShift.defaults
			)
	})()